alerters/alerters-plugins/alerters-gnocchi/src/main/java/org/hawkular/alerter/gnocchi/GnocchiQuery.java

```diff
@@ -0,0 +1,403 @@
+        this.properties = properties == null ? new HashMap<>() : new HashMap<>(properties);
+        this.alerts = alerts;
+        String user = null;
+        String password = null;
+        if (trigger != null) {
+            baseUrl = trigger.getContext().get(GnocchiAlerter.URL);
+            user = trigger.getContext().get(GnocchiAlerter.USER);
+            password = trigger.getContext().get(GnocchiAlerter.PASSWORD);
+            metricIds = trigger.getContext().get(METRIC_IDS);
+            metricNames = trigger.getContext().get(METRIC_NAMES);
+            metricNamesRegexp = trigger.getContext().get(METRIC_NAMES_REGEXP);
+            metricResourceQuery = trigger.getContext().get(METRIC_RESOURCE_QUERY);
+            metricAggregation = trigger.getContext().get(METRIC_AGGREGATION);
+            interval = trigger.getContext().get(INTERVAL) == null ? INTERVAL_DEFAULT : trigger.getContext().get(INTERVAL);
+            granularity = trigger.getContext().get(METRIC_GRANULARITY) == null ? METRIC_GRANULARITY_DEFAULT : trigger.getContext().get(METRIC_GRANULARITY);
+        }
+        if (isEmpty(baseUrl)) {
+            baseUrl = this.properties.getOrDefault(GnocchiAlerter.URL, GnocchiAlerter.GNOCCHI_URL_DEFAULT);
+        }
+        if (isEmpty(user)) {
+            user = this.properties.getOrDefault(GnocchiAlerter.USER, GnocchiAlerter.GNOCCHI_USER_DEFAULT);
+        }
+        if (isEmpty(password)) {
+            password = this.properties.getOrDefault(GnocchiAlerter.PASSWORD, GnocchiAlerter.GNOCCHI_PASSWORD_DEFAULT);
+        }
+        basicAuth = new StringBuilder("Basic ")
+                .append(new String(Base64.getEncoder().encode(new StringBuilder()
+                        .append(user)
+                        .append(":")
+                        .append(password)
+                        .toString()
+                        .getBytes())))
+                .toString();
+        searchMetrics();
+        processMetricAggregation();
+        buildGnocchiQueries();
+    }
+
+    public void searchMetrics() {
+        // Case 1, fixed list of metric ids, so, no need to query names or anything
+        if (!isEmpty(metricIds)) {
+            String[] gnocchiIds = metricIds.split(",");
+            for (int i = 0; i < gnocchiIds.length; i++) {
+                nameIdMetrics.put(gnocchiIds[i].trim(), gnocchiIds[i].trim());
+            }
+            log.debugf("Gnocchi Metrics configured %s", nameIdMetrics);
+            return;
+        }
+
+        // Case 2, fixed list of metric names, a query to get gnocchi ids is necessary
+        if (!isEmpty(metricNames) && isEmpty(metricResourceQuery)) {
+            String[] splitMetricNames = metricNames.split(",");
+            List<String> gnocchiNames = new ArrayList<>();
+            for (int i = 0; i < splitMetricNames.length; i++) {
+                gnocchiNames.add(splitMetricNames[i].trim());
+            }
+            List rawAllMetrics = getAllMetrics();
```

Comment:
I'd suggest passing at least `?name=metric_name` to the query to limit the output. If you got millions of metrics it's gonna be a problem. :)

Reply:
+1, it is the default case when it is not defined, an alternative is to throw some error message forcing user to define it, but yes, idea is that user can define metrics in several ways
https://github.com/hawkular/hawkular-alerts/pull/467/files#diff-f7efcfbcd8455ebf114f6e51ced5131bR111
i.e.
metric.ids, or
metric.names, or
metric.names.regexp, and
metric.resource.query, to even narrow the potential metrics to search.

Also, this is performed only once, to resolve internally metric name -> metric id which will be used for query measures.